feat(hogql): select from s3

mariusandra opened the pull request:
## Problem

We need support `select from function()` for two cases:

- selecting from `s3(user uploaded large file)` (data warehouse is working on this CC @timgl)
- selecting from the numbers table (blocking moving trends to a hogql insight)

## Changes

Implements the first of the two, and leaves some building blocks for the second.

I did not expose the `s3` method directly, instead created a `S3Table` class that can be used as a hogql database table.

I also had to change to `readonly=2` from `=1` to [allow temporary tables](https://clickhouse.com/docs/en/operations/settings/permissions-for-queries). That flag also gives access to `SET something=1` queries, but HogQL can only create "SELECT" queries anyway, so we're good. If someone finds a way to inject arbitrary text into the generated SQL, then we likely have bigger issues than this.

![2023-06-14 22 04 23](https://github.com/PostHog/posthog/assets/53387/8ba70157-128c-4910-ae0b-6e02593684b7)


## How did you test this code?

Wrote some tests. Added a hidden `aapl_stock` table that you can select from to try it out. It's from the ClickHouse docs.

mariusandra committed:
add a few types

mariusandra committed:
context value adding

mariusandra committed:
first draft of a system that sort of works

mariusandra committed:
hacks to get it working

mariusandra committed:
fix alias

timgl commented on posthog/hogql/database/s3_table.py:
```diff
@@ -0,0 +1,29 @@
+from posthog.hogql.escape_sql import escape_hogql_identifier
+
+
+class S3Table(TableFunction):
```
I called this `ExternalTable`, as we'll support GCP/Azure and potentially others. Don't feel very strongly about this.

timgl commented on docker/clickhouse/config.xml:
```diff
@@ -677,6 +677,7 @@
             Regexps are not aligned: don't forget to add ^ and $. Also don't forget to escape dot (.) metacharacter
             (forgetting to do so is a common source of error).
         -->
+        <host>datasets-documentation.s3.eu-west-3.amazonaws.com</host>
```
```suggestion
            <host_regexp>^s3.*\.amazonaws\.com$</host_regexp>
            <host_regexp>^.*\.googleapis\.com$</host_regexp>
```

Makes it consistent with prod

mariusandra committed:
simplify a lot

mariusandra committed:
set readonly to 2

mariusandra committed:
Update docker/clickhouse/config.xml

Co-authored-by: timgl <tim@glsr.nl>

mariusandra committed:
Merge branch 'select-from-s3' of github.com:PostHog/posthog into select-from-s3

mariusandra committed:
make it pass our list of urls

mariusandra committed:
better aliasing

mariusandra committed:
remove flimsy thing

mariusandra committed:
different type of input

mariusandra committed:
dead code

mariusandra committed:
refactor aapl_stock

mariusandra committed:
own it

mariusandra committed:
Merge branch 'master' into select-from-s3

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
more cleanup

mariusandra committed:
Merge branch 'select-from-s3' of github.com:PostHog/posthog into select-from-s3

mariusandra commented:
I did plenty of changes since the last review, and this should be good to go now, assuming CI agrees.

@timgl btw the `s3.*\.amazonaws\.com` enabled host didn't match the `https://datasets-documentation.s3.eu-west-3.amazonaws.com/aapl_stock.csv` URL I was using. I moved the bucked to the path and didn't change anything here, but perhaps `.*\.s3.*\.amazonaws\.com` needs to be added to the list as well?

mariusandra committed:
Merge branch 'master' into select-from-s3

timgl commented:
Ah yep I'll add that to my PR and update prod

github-actions[bot] committed:
Update query snapshots

Twixes reviewed (commented):
Looks great to me
